service/lib/agama/dbus/storage/proposal_settings_conversion/from_dbus.rb

```diff
@@ -0,0 +1,165 @@
+          private_constant :CONVERSIONS
+
+          # @param target [Agama::Storage::ProposalSettings]
+          # @param value [String]
+          def boot_device_conversion(target, value)
+            target.boot_device = value
+          end
+
+          # @param target [Agama::Storage::ProposalSettings]
+          # @param value [Boolean]
+          def lvm_conversion(target, value)
+            target.lvm.enabled = value
+          end
+
+          # @param target [Agama::Storage::ProposalSettings]
+          # @param value [Array<String>]
+          def system_vg_devices_conversion(target, value)
+            target.lvm.system_vg_devices = value
+          end
+
+          # @param target [Agama::Storage::ProposalSettings]
+          # @param value [String]
+          def encryption_password_conversion(target, value)
+            target.encryption.password = value.empty? ? nil : value
+          end
+
+          # @param target [Agama::Storage::ProposalSettings]
+          # @param value [String]
+          def encryption_method_conversion(target, value)
+            method = Y2Storage::EncryptionMethod.find(value.to_sym)
+            return unless method
+
+            target.encryption.method = method
+          end
+
+          # @param target [Agama::Storage::ProposalSettings]
+          # @param value [String]
+          def encryption_pbkd_function_conversion(target, value)
+            function = Y2Storage::PbkdFunction.find(value)
+            return unless function
+
+            target.encryption.pbkd_function = function
+          end
+
+          # @param target [Agama::Storage::ProposalSettings]
+          # @param value [String]
+          def space_policy_conversion(target, value)
+            policy = value.to_sym
+            return unless Agama::Storage::SpaceSettings.policies.include?(policy)
+
+            target.space.policy = policy
+          end
+
+          # @param target [Agama::Storage::ProposalSettings]
+          # @param value [Hash]
+          def space_actions_conversion(target, value)
+            target.space.actions = value
+          end
+
+          # @param target [Agama::Storage::ProposalSettings]
+          # @param value [Array<Hash>]
+          def volumes_conversion(target, value)
+            # Keep default volumes if no volumes are given
+            return if value.empty?
+
+            required_volumes = target.volumes.select { |v| v.outline.required? }
+            volumes = value.map { |v| VolumeConversion.from_dbus(v, config: config) }
+
+            target.volumes = volumes + missing_volumes(required_volumes, volumes)
+          end
+
+          # Missing required volumes
+          #
+          # @param required_volumes [Array<Agama::Storage::Volume>]
+          # @param volumes [Array<Agama::Storage::Volume>]
+          #
+          # @return [Array<Agama::Storage::Volume>]
+          def missing_volumes(required_volumes, volumes)
+            mount_paths = volumes.map(&:mount_path)
```

Comment:
just small efficiency hint. Set allows much faster searching so with `require "set" you can use  `mount_paths = volumes.map(&:mount_path).to_set` and then include is much faster

Reply:
Damn, set checks for object equality and not for its value, so it will probably not work here :( https://apidock.com/ruby/v2_5_5/Set/include%3F So lets keep it as it is, I think we should not have too much volumes.

Reply:
my quick test in irb shows that it should work, but who knows :)